### ISSUE TITLE
pgo scaledown does not work if postgresql is configured with tls/ssl encryption

### ISSUE THREAD

**Issue by yuanlinios** [CrunchyData/postgres-operator#1180]
**Describe the bug**
`pgo scaledown` does not work if pg cluster is configured with tls/ssl encryption (encryption only, no tls/ssl authN)

**To Reproduce**
Customized `pg_hba.conf`
```
local    all          all                               trust
host     all          all           127.0.0.1/32        md5
hostssl  all          all           0.0.0.0/0           md5
hostssl  replication  primaryuser   0.0.0.0/0           md5
```

Create pg cluster with custom config
```
oc create -n pg-proj configmap pg-custom-config \
    --from-file=configs/server.crt \
    --from-file=configs/server.key \
    --from-file=configs/pg_hba.conf \
    --from-file=configs/postgresql.conf

pgo create cluster devcluster --replica-count=1 --custom-config pg-custom-config -n pg-proj
```

Try to run `pgo scaldown` coomand
```
pgo scaledown devcluster -n pg-proj --query
Error: pq: no pg_hba.conf entry for host "10.9.3.4", user "postgres", database "postgres", SSL off
```

**Expected behavior**
`pgo scaledown` command should be executed successfully

**Please tell us about your environment:**

* Container Image Tag: crunchydata/crunchy-postgres:centos7-12.1-4.1.1
* PostgreSQL Version: 12.1
* Platform (Docker, Kubernetes, OpenShift): Openshift
* Platform Version: OCP 3.11

**Additional context**
The documentation mentions only following commands do not work with ssl enabled in pg cluster
```
pgo test
pgo load
pgo apply
```

**Comment by jkatz** [comment]
Hi @yuanlinios,

Thanks for the report!

The behavior of `pgo scaledown` has changed in the upcoming v4.2.0 release (which you can read about here: https://github.com/CrunchyData/postgres-operator/releases/tag/v4.2.0-beta.1) where it does not perform a query directly to the PostgreSQL database in order to get the information as part of the `--query` flag. As such, this issue is resolved in that release as it does not matter if PostgreSQL has TLS enabled

Similarly, in `v4.2.0`, `pgo test` no longer connects directly to the PostgreSQL database and as such the issue is resolved there.

As the nature of these changes qualify as new features, they will not be backpatched to `REL_4_1`.

The issues you mention with TLS regarding `pgo load` and `pgo apply` at present would still be issues in `v4.2.0`.  I will see if there is something we can do about fixing it for the upcoming release, though I will mention we're very late in the release cycle and cannot promise anything.

Thanks!

**Comment by jkatz** [comment]
@yuanlinios I've addressed the `pgo apply` issue in PR #1182

**Comment by jkatz** [comment]
@yuanlinios Looking at your example, #1182 may also solve the issue you are seeing in `pgo load` as well, as `pgo load` has the option to apply policies before the load is executed. Within the load script itself, the connection is attempted to the PostgreSQL database via the `psql` command, which does not make any references to `sslmode` and therefore uses the default of `prefer`. With the fix in #1182 (and perhaps without if you are not using `--policies`) this should be able to connect with this setting as is, but I would be curious to hear about the behavior you're seeing.

**Comment by yuanlinios** [comment]
Hello @jkatz 

Thanks for your quick reply! I am looking forward to the 4.2 release.

For `pgo load` without `--policies`, I run the example under `examples/load` in my environment and it works OK with TLS/SSL enabled in pg cluster.


During my test, I noticed even more features broken with TLS/SSL enabled in pg cluster
```
pgo df
pgo failover --query
pgo create/update/delete user
```
Are these going to be fixed like `pgo scaledown` in the coming release? Some of them are important features for daily operation. And for security reasons, I believe TLS encryption is a common requirement in public cloud environment.

**Comment by jkatz** [comment]
Hi @yuanlinios 

Thanks for the feedback, this is incredibly helpful! I agree that TLS is very important for your use case, and as someone who used to run the PostgreSQL Operator before programming on it, I could not agree more. In upcoming releases, you will see more to ensure we are safely and properly supporting TLS.

Another question for you: when you set up your TLS enabled PostgreSQL instances, how are you setting up your PKI? Do you have a trusted CA that you are importing? Are you leveraging the CA that Kubernetes provides?

To answer your questions:

1. `pgo failover --query` follows the same fix for `pgo scaledown --query`, this is handled in 4.2
1. There are some tweaks I would like to make to `pgo df` to make it more helpful for day-to-day use, but I won't be able to do that for 4.2.
1. Similarly for the `user` series of commands. We started adding a sidecar in 4.2 called `crunchyadm` that is available to perform secure administrative functions against a PostgreSQL database, and in an upcoming release you will see how it will help with the `user` series of commands.

Thanks!

**Comment by yuanlinios** [comment]
Hello @jkatz 

Thanks for your reply! Here is some info about my PKI setup

I have a typical 2-tiered private PKI in my lab environment:  root CA + issuing CA. Root CA cert is trusted in all VM instances during their initial provisioning. 

For ocp, I use the default self-provisioned ocp CA and etcd CA, and provide private issuing CA signed certificates for routes and external master api access.

For TLS setup in containerized applications (e.g. postgres), I also use the private issuing CA to sign certificates

Feel free to let me know if you want more details

**Comment by jkatz** [comment]
Hi @yuanlinios,

Thanks for the feedback. This is certainly incredibly helpful and should help guide some of our future decisions to better support TLS.

I also wanted to let you know that [`v4.2.0`](https://github.com/CrunchyData/postgres-operator/releases/tag/v4.2.0) is now available.

**Comment by yuanlinios** [comment]
Hello @jkatz 

Sorry for the delay. I am trying the latest 4.2.1 release these days, and I can confirm `pgo failover --query` and `pgo scaledown --query` are now working with TLS enabled pgcluster.

I will try more scenarios

**Comment by jkatz** [comment]
Hi @yuanlinios 

Awesome. I'm actively working through this + #901 which will take care of all of the others. Making fairly good progress too (as you might see from the above commits 😄)

If you're curious to track, my WIP branch is here: https://github.com/jkatz/postgres-operator/tree/tls-harden

**Comment by yuanlinios** [comment]
Hello @jkatz 

I read through #901, and tried 

Feature 1 in #901 currently breaks some features in an TLS encryption enabled cluster.
-  `pgo df`
- `pgo create user`
- `pgo create pgbouncer`

2/3/4 could be problems if a user configures his postgres cluster with TLS authN because currently operator/pgo-load/pgdump use password based authentication. However different postgres cluster may be configured with different pg_hba.conf authN options, it might be infeasible to adapt these components to all different postgres cluster authN configurations

**Comment by jkatz** [comment]
Hi @yuanlinios 

In the branch I linked to in the previous comment, I've already fixed it for `pgo df` and `pgo create pgbouncer` (and beyond). The last set is `pgo * user` which I am starting on this week.

Please don't read into the text of #901 as that is not the methodology I am necessarily using, rather I am just referencing the ticket as it discusses support for TLS only. Look at the branch, particularly:

https://github.com/jkatz/postgres-operator/commits/tls-harden

Thanks,

Jonathan

**Comment by jkatz** [comment]
Hi @yuanlinios,

This is now merged. All of the Operator actions now work in a PostgreSQL deployment that is TLS-only. This will be available in 4.3.0

Thanks!
